### PR TITLE
Install OSX command line tools

### PR DESCRIPTION
**Pull request by elliotweiser**
Summary of Changes
==================

Since the OS X Command Line Tools are a hard dependency for Homebrew, this
role will now make sure that they are installed. If they are already
installed, those steps will be skipped. This makes using this role easier
to use when provisioning MacOS, especially when the destination host is not
the same as the control node.

Other related changes:

* `meta/main.yml`: specifies platform `MacOSX` instead of `GenericUNIX`
* `.travis.yml`: Fix homebrew uninstall process-- the old way hangs.
* `tasks/main.yml`: Ensure `root` is the owner of `homebrew_prefix` (`/usr/local`)

### CODE REVIEW

File: tests/test_default.py [hunk 1]
```diff
@@ -0,0 +1,21 @@
+def test_brew_taps(Command):
+    c = Command('/usr/local/bin/brew tap --list')
+    assert c.rc == 0
+    tapped = c.stdout.split()
+    assert 'homebrew/core' in tapped
```
Comment (geerlingguy) [inline tests/test_default.py]: Forgive me if I'm being dense... but wouldn't this just be testing that Ansible's core modules are doing what they say they do? The role just tells Ansible to verify the taps are tapped and the packages are installed—if that was broken, then the role's normal tests (run it, run it again) would fail.

I guess I'm failing to see why I'd want to add a second layer of tests to this role which simply asserts that the tests Ansible itself runs during the playbook run are passing? Like checking if `yes == yes` actually equals `yes == yes`... if it doesn't, that's an upstream issue and I'd start looking at an alternative solution besides Ansible if I can't trust it to at least verify something is installed or not :/

Comment (elliotweiser) [inline tests/test_default.py]: > Wouldn't this just be testing that Ansible's core modules are doing what they say they do?

Not quite-- it's for testing that the _role_ does what it says it does. Take the following scenarios:

1. A developer deletes the task that installs homebrew packages.
1. A developer changes the default cask applications such that e.g. `google-chrome` is installed rather than `firefox`.

Simply running the test playbook and the idempotence test would not catch these issues. 

Testinfra verifies that the desired state on the target host(s) has been attained. Think of it like functional tests alongside your application code. 

The idempotence test is still essential and I don't intend to downplay its significance. IMO, "run it" and "run it again" is more of smoke test.

I hope that answers your question 😄

**Comment by elliotweiser** [comment]
There are still a few kinks to work out. Whether the control node and the target host are the same makes a difference. I'll need to make sure that the behavior over SSH matches the local behavior.

**Comment by elliotweiser** [comment]
All fixed. I can think of some additional (potential) changes, but I think they can be saved for a separate PR. Thanks for the reviews!

**Comment by elliotweiser** [comment]
Fixes #45 ; Installing the OS X Command Line Tool makes the `git` binary usable.

File: tests/test_default.py (hunk 1 above)
Comment (geerlingguy) [inline tests/test_default.py]: @elliotweiser - But a developer shouldn't ever touch the roles themselves—they should be managed via `requirements.yml` or in some other way as a dependency. It's one case when designing software, another when dealing with infra. I'm working on a longer form post on why I believe this to be the case, but the gist of it is: since I maintain all these roles, and keep things as simple/efficient as possible, I don't want to maintain a separate set of tests that forms the 'contract' for the roles. I know what I want them to do, and that's enough for me :)

If the roles were maintained by a larger team, or if they were large enough that I couldn't fit the entire problem space in my head when working on them, I would be more prone to adding functional/system tests _on top of_ the built-in test that running the role provides, but for this role (and almost all the rest of my Ansible roles), they're simple enough that I could recite the list of tasks and includes almost by heart :)

Tests require maintenance, even in simpler cases, and for a guy maintaining over 150 projects on GitHub, and dozens more elsewhere, I only add tests where they are essential to my ability to continue maintaining the roles over a long period of time.

Anyways, I guess what I'm saying here is that I don't disagree with you, but in this case I don't want to merge in any extra test layers for this role (e.g. molecule, plus the extra test kitchen stuff).

And I would still like to have this PR split up further—can you break each change into it's own PR?

Comment (elliotweiser) [inline tests/test_default.py]: Sure, I'll split up the changes into separate logical PRs. It started smaller but gradually got away from me...

Perhaps I misspoke. The "developers" in those scenarios were meant to be potential contributors to _this_ project. I agree that users should not be attempting to modify the code for this role in any other context. What I see happening is that the Travis build passes, and only code-review stands in the way of introducing potentially harmful/breaking changes. 

In fact, that speaks pretty well to what I'm going to say next...

What I take away from your last response is that you have concerns about the maintainability of these changes. I certainly appreciate how challenging it can be to manage that many projects. However, I will assert (to the continued annoyance of my friends/coworkers) that contract-defining tests, even for infrastructure automation, actually makes projects easier to maintain. The cost of writing/maintaining those tests is vastly outweighed by the benefit of having them. There is no doubt a longer a conversation we could have on that topic (happy to take it offline). 

Let me know if you'd like any assistance with maintaining this project. Feel free to add me as a collaborator! 😄 

FWIW, the `molecule` portion makes local testing _really_ easy.

File: README.md [hunk 2]
```diff
@@ -59,7 +59,7 @@ The directory where your Brewfile is located.
 
 ## Dependencies
 
-None.
+* [elliotweiser.osx-command-line-tools](https://galaxy.ansible.com/elliotweiser/osx-command-line-tools/)
```
Comment (geerlingguy) [inline README.md]: What would this dependency be used for?

Comment (geerlingguy) [inline README.md]: Ah... to actually install the CLI tools. In this case, is there a missing commit that uses the role?

Comment (geerlingguy) [inline README.md]: Also, just as an aside... isn't it technically against Apples terms to host OS X boxes for VirtualBox? https://github.com/elliotweiser/ansible-osx-command-line-tools/blob/master/molecule.yml#L13-L22 shows various boxes for different OS releases. I love the idea, but the reason I've never put any of my own boxes on Atlas is the legality...

File: tasks/main.yml [hunk 3]
```diff
@@ -10,9 +10,11 @@
   file:
     path: "{{ homebrew_prefix }}"
     mode: 0775
+    owner: root
     group: admin
     state: directory
   become: yes
+  become_user: root
```
Comment (geerlingguy) [inline tasks/main.yml]: `become` usually defaults to `root`, though—or if globally configured otherwise, that different user... I'd rather either leave this as-is, or switch it to a configurable user, defaulted to `root`.

File: README.md (hunk 2 above)
Comment (elliotweiser) [inline README.md]: By defining the dependency in `meta/main.yml`, the dependency role is executed first. 

I believe Apple allows you to run up to VMs (virtualbox or otherwise) as long as it they _run on Mac hardware_. I'll see if I can get my hands on the actual legalese...

File: tasks/main.yml (hunk 3 above)
Comment (elliotweiser) [inline tasks/main.yml]: That's fine. I had trouble testing on a remote machine over ssh, but I think that can be addressed later.

File: README.md (hunk 2 above)
Comment (elliotweiser) [inline README.md]: Just to clarify, there is no missing commit. This change already ensures that the CLTs are installed because the homebrew role executes its dependency prior to executing its own tasks.

Also, I found the SLAs:

http://images.apple.com/legal/sla/docs/macOS1012.pdf
http://images.apple.com/legal/sla/docs/OSX1011.pdf
http://images.apple.com/legal/sla/docs/OSX10103.pdf

Section 2.B.iii confirms what I said.

Comment (geerlingguy) [inline README.md]: > (iii) to install, use and run up to two (2) additional copies or instances of the Apple Software within virtual operating system environments on each Mac Computer you own or control that is already running the Apple Software, for purposes of: (a) software development; (b) testing during software development; (c) using macOS Server; or (d) personal, non-commercial use.

Hah, I think every month I install about 10 or so extra instances :P

File: meta/main.yml [hunk 4]
```diff
@@ -1,5 +1,7 @@
 ---
-dependencies: []
+dependencies:
+  - role: elliotweiser.osx-command-line-tools
+    version: 1.1.0
```
Comment (geerlingguy) [inline meta/main.yml]: Locking into a version does help stability somewhat... but also means it's another thing to maintain as the upstream role is improved/fixed. Since there's no concept of a 'major version' I can lock to, I would rather not favor a particular version, but instead keep this role working with the upstream and suggest the latest (e.g. just `- elliotweiser.osx-command-line-tools`).

Comment (elliotweiser) [inline meta/main.yml]: Sure, I'll unpin the version.

**Comment by elliotweiser** [comment]
Referencing my earlier comment from https://github.com/geerlingguy/ansible-role-homebrew/pull/53#discussion_r93993366

TL;DR: I'd be happy to assist with the maintenance of this project. I have the time. Feel free to add me as a collaborator.

**Comment by elliotweiser** [comment]
@geerlingguy I removed the version from `requirements.yml` for consistency.

**Comment by elliotweiser** [comment]
Updated the branch and resolved conflicts. I think is ready to go. Let me know if there are any other improvements to be made.

**Comment by geerlingguy** [comment]
Thanks! I have this on my short list of things to poke around more—just haven't had the time to give it a final review; and I really wanted to get that other bug (https://github.com/geerlingguy/ansible-role-homebrew/pull/55) fixed ASAP, so sorry about forcing the rebase :(

**Comment by elliotweiser** [comment]
No worries! I ran into similar issues myself, so it was good to nip that in the bud. In the interest of knowledge-sharing, I'll link my workaround on the related PR(s).

**Comment by geerlingguy** [comment]
+1
